cpp/include/cuspatial/detail/utility/linestring.cuh

```diff
@@ -130,5 +137,66 @@ __forceinline__ T __device__ squared_segment_distance(vec_2d<T> const& a,
   return segment_distance_no_intersect_or_colinear(a, b, c, d);
 }
 
+/**
+ * @internal
+ * @brief Given two collinear or parallel segments, return their potential overlapping segment
+ *
+ * @return optional end points of overlapping segment
+ */
+template <typename T>
+__forceinline__ thrust::optional<segment<T>> __device__ collinear_or_parallel_overlapping_segments(
+  vec_2d<T> const& a, vec_2d<T> const& b, vec_2d<T> const& c, vec_2d<T> const& d)
+{
+  auto ab = b - a;
+  auto ac = c - a;
+
+  // Parallel
+  if (det(ab, ac) != 0) return thrust::nullopt;
+
+  // Must be on the same line, test if intersect
+  if ((a < c && c < b) || (a < d && d < b)) {
+    // Compute smallest interval between the segments
+    auto a_ = a, b_ = b, c_ = c, d_ = d;
+    if (d_ < c_) thrust::swap(c_, d_);
+    if (b_ < a_) thrust::swap(a_, b_);
+    return segment<T>{a_ > c_ ? a_ : c_, b_ < d_ ? b_ : d_};
+  }
+
+  return thrust::nullopt;
+}
+
+/**
+ * @internal
+ * @brief Primitive to compute intersections between two segments
+ * Two segments can intersect at a point, overlap at a segment, or does not have common set.
+ *
```

Comment:
Might want a note that the return value is *either* a point or a segment -- cannot be both. I wonder if this means that we could use `std::variant` as the return type?

Reply:
Yeah this also comes to me as well. Will look at this tomorrow.

Reply:
But variant cannot represent nullset... Perhaps `std::optional<std::variant<vec_2d<T>, segment<T>>>`?

Reply:
`std::vairant` isn't supported in device code yet. https://github.com/NVIDIA/libcudacxx/issues/162

Reply:
OK, proceed with the current approach then. :)